api/pkg/controller/rancherinit/controller.go

```
@@ -0,0 +1,462 @@
+	defer cancel()
+
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+	if request.Name != resources.RancherStatefulSetName {
+		return reconcile.Result{}, nil
+	}
+	statefulSet := &appsv1.StatefulSet{}
+	if err := r.Get(ctx, request.NamespacedName, statefulSet); err != nil {
+		if kubeapierrors.IsNotFound(err) {
+			log.Errorw("Couldn't find statefulSet", zap.Error(err))
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// all done.
+	if statefulSet.Annotations["kubermatic.io/rancher-server-cluster-imported"] == "true" {
+		return reconcile.Result{}, nil
+	}
+
+	result, err := r.reconcile(ctx, log, statefulSet)
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, statefulSet *appsv1.StatefulSet) (*reconcile.Result, error) {
+	if statefulSet.DeletionTimestamp != nil {
+		return nil, nil
+	}
+	// wait for statefulSet to be ready
+	if statefulSet.Status.ReadyReplicas == 0 {
+		return &reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+	if statefulSet.Annotations == nil {
+		statefulSet.Annotations = make(map[string]string)
+	}
+	if statefulSet.Annotations["kubermatic.io/rancher-server-ready"] != "true" { // if we have ready replicas, we should be ready
+		statefulSet.Annotations["kubermatic.io/rancher-server-ready"] = "true"
+		if err := r.Update(ctx, statefulSet); err != nil {
+			if kubeapierrors.IsConflict(err) {
+				return &reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+			}
+			log.Errorw("failed to update rancher statefulSet", zap.Error(err))
+			return nil, err
+		}
+	}
+	// initialize rancher statefulSet
+	if statefulSet.Annotations["kubermatic.io/rancher-server-initialized"] != "true" {
+		if err := r.initRancherServer(ctx, log, statefulSet); err != nil {
+			log.Errorw("failed to initialize Rancher Server", zap.Error(err))
+			return &reconcile.Result{RequeueAfter: 10 * time.Second}, nil
```

This should return an error instead of a `RequeueAfter` so the exponential backoff is used